core/testutils/src/main/java/com/mongodb/stitch/core/testutils/sync/SyncIntTestProxy.kt

```diff
@@ -0,0 +1,1361 @@
+            sem.acquire()
+            streamAndSync()
+
+            // it should not have updated the local doc, as the local doc should be frozen
+            assertEquals(
+                withoutId(expectedDoc),
+                withoutSyncVersion(withoutId(testSync.find(Document("_id", result.insertedId)).first()!!)))
+
+            // update the local doc. this should unfreeze the config
+            testSync.updateOneById(result.insertedId, Document("\$set", Document("no", "op")))
+
+            streamAndSync()
+
+            // this should still be the remote doc since remote wins
+            assertEquals(
+                withoutId(nextDoc),
+                withoutSyncVersion(withoutId(testSync.find(Document("_id", result.insertedId)).first()!!)))
+
+            // update the doc remotely
+            val lastDoc = Document("good night", "computer")
+
+            sem = watchForEvents(syncTestRunner.namespace)
+            remoteColl.updateOne(
+                Document("_id", result.insertedId),
+                withNewSyncVersion(lastDoc)
+            )
+            sem.acquire()
+
+            // now that we're sync'd and unfrozen, it should be reflected locally
+            // TODO: STITCH-1958 Possible race condition here for update listening
+            streamAndSync()
+
+            assertEquals(
+                withoutId(lastDoc),
+                withoutSyncVersion(
+                    withoutId(testSync.find(Document("_id", result.insertedId)).first()!!)))
+        }
+    }
+
+    @Test
+    fun testConfigure() {
+        val coll = syncTestRunner.syncMethods()
+        val remoteColl = syncTestRunner.remoteMethods()
+
+        // insert a documnet locally
+        val docToInsert = Document("hello", "world")
+        val insertedId = coll.insertOneAndSync(docToInsert).insertedId
+
+        var hasConflictHandlerBeenInvoked = false
+        var hasChangeEventListenerBeenInvoked = false
+
+        // configure Sync, each entry with flags checking
+        // that the listeners/handlers have been called
+        val changeEventListenerSemaphore = Semaphore(0)
+        coll.configure(
+            ConflictHandler { _: BsonValue, _: ChangeEvent<Document>, remoteEvent: ChangeEvent<Document> ->
+                hasConflictHandlerBeenInvoked = true
+                assertEquals(remoteEvent.fullDocument["fly"], "away")
+                remoteEvent.fullDocument
+            },
+            ChangeEventListener { _: BsonValue, _: ChangeEvent<Document> ->
+                hasChangeEventListenerBeenInvoked = true
+                changeEventListenerSemaphore.release()
+            },
+            ErrorListener { _, _ -> }
+        )
+
+        // insert a document remotely, and wait for the event to store
+        val sem = watchForEvents(syncTestRunner.namespace)
+        remoteColl.insertOne(Document("_id", insertedId).append("fly", "away"))
+        sem.tryAcquire(3, TimeUnit.SECONDS)
```

Comment:
[mahor] none of these tests that previously had just a acquire should have try acquire now. That encourages races to occur. If 3 seconds pass and nothing happens, this proceeds.